common/src/stack/pylib/stack/kvm.py

```diff
@@ -0,0 +1,376 @@
+# @copyright@ # Copyright (c) 2006 - 2018 Teradata
+# All rights reserved. Stacki(r) v5.x stacki.com
+# https://github.com/Teradata/stacki/blob/master/LICENSE.txt
+# @copyright@
+#
+
+from pexpect import pxssh
+import os
+import libvirt
+import tarfile
+from libvirt import libvirtError
+from stack.util import _exec
+import getpass
+from pathlib import Path
+
+# Libvirt has a weird quirk where it prints
+# error messages even if they are handled via
+# exceptions, code taken from:
+# stackoverflow.com/questions/45541725/avoiding-console-prints-by-libvirt-qemu-python-apis
+def libvirt_callback(userdata, err):
+    pass
+
+class VmException(Exception):
+	pass
+
+class VM:
+
+	"""
+	Class to interface to a kvm hypervisor via
+	the libvirt bindings. Can create and remove vm's,
+	get the current state of guests, and manage
+	storage pools or images.
+	"""
+
+	def __init__(self, host, privkey):
+		libvirt.registerErrorHandler(f=libvirt_callback, ctx=None)
+		self.hostname = host
+		self.privkey = privkey
+		self.sshCopyID()
+		self.kvm = self.connect()
+
+	def doRemoteCommand(self, user, commands):
+		""""
+		Try to execute a command on a remote host via ssh.
+		If there is an ssh  key use that, otherwise prompt for a
+		password and send a command.
+		"""
+
+		try:
+			s = pxssh.pxssh()
+
+			if self.privkey:
+				s.login(self.hostname, user, ssh_key = self.privkey)
+			else:
+				password = getpass.getpass('root password for "%s": ' % self.hostname)
+				s.login(self.hostname, user, password)
+
+			for cmd in commands:
+				s.sendline(cmd)
+				s.prompt()
+
+			s.logout()
+
+		except pxssh.ExceptionPxssh as e:
+			return False
+
+		return True
+
+	def sshCopyID(self):
+		"""
+		Initialize the ssh connection between a frontend and a KVM system
+
+		Raises a VmException if the public key on the frontend could
+		not be found
+		"""
+
+		pubkey = '%s.pub' % self.privkey
+		if not os.path.exists(pubkey):
+			raise VmExcpetion(f'Failed to find public key to contact hypervisor at {pubkey}')
+
+		try:
+			f = open(pubkey)
+			stacki_public_key = f.read()
+			f.close
+		except IOError:
+			raise VmExcpetion(f'Failed to read public key to contact hypervisor at {pubkey}')
+
+		if not self.doRemoteCommand('root', []):
+			commands = [ 'echo "%s" >> /root/.ssh/authorized_keys' % stacki_public_key ]
```

Comment:
Use an f-string here, since you use f-strings most other places. Also, you don't need the temp `commands` variable.